lib/logstash/outputs/elasticsearch_java/protocol.rb

```diff
@@ -0,0 +1,248 @@
+            end
+          end
+
+          def bulk(actions)
+            # Actions an array of [ action, action_metadata, source ]
+            prep = @client.prepareBulk
+            actions.each do |action, args, source|
+              prep.add(build_request(action, args, source))
+            end
+            response = prep.execute.actionGet()
+
+            self.class.normalize_bulk_response(response)
+          end # def bulk
+
+          def build_request(action, args, source)
+            case action
+              when "index"
+                request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
+                request.id(args[:_id]) if args[:_id]
+                request.routing(args[:_routing]) if args[:_routing]
+                request.source(source)
+              when "delete"
+                request = org.elasticsearch.action.delete.DeleteRequest.new(args[:_index])
+                request.id(args[:_id])
+                request.routing(args[:_routing]) if args[:_routing]
+              when "create"
+                request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
+                request.id(args[:_id]) if args[:_id]
+                request.routing(args[:_routing]) if args[:_routing]
+                request.source(source)
+                request.opType("create")
+              when "create_unless_exists"
+                unless args[:_id].nil?
+                  request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
+                  request.id(args[:_id])
+                  request.routing(args[:_routing]) if args[:_routing]
+                  request.source(source)
+                  request.opType("create")
+                else
+                  raise(LogStash::ConfigurationError, "Specifying action => 'create_unless_exists' without a document '_id' is not supported.")
+                end
+              when "update"
+                unless args[:_id].nil?
+                  request = org.elasticsearch.action.update.UpdateRequest.new(args[:_index], args[:_type], args[:_id])
+                  request.routing(args[:_routing]) if args[:_routing]
+                  request.doc(source)
+                  if @options[:doc_as_upsert]
+                    request.docAsUpsert(true)
+                  else
+                    request.upsert(args[:_upsert]) if args[:_upsert]
+                  end
+                else
+                  raise(LogStash::ConfigurationError, "Specifying action => 'update' without a document '_id' is not supported.")
+                end
+              else
+                raise(LogStash::ConfigurationError, "action => '#{action_name}' is not currently supported.")
+            end # case action
+
+            request.type(args[:_type]) if args[:_type]
+            return request
+          end # def build_request
+
+          def template_exists?(name)
+            return !client.admin().indices().
```

Comment:
this method uses foo.method().method() and the next doesn't use parens foo.method.method. Remove the parens maybe? Thoughts?